screens/locationData.js

```diff
@@ -0,0 +1,179 @@
+    const {
+      failedOpen,
+      numberUploading,
+    } = this.state;
+
+    const iconName = failedOpen ? 'chevron-small-down' : 'chevron-small-right';
+    const disabled = stats.offline === 0 && stats.failed === 0;
+
+    return (
+      <View style={styles.container}>
+        <View>
+          <TouchableOpacity style={styles.button} disabled={disabled} onPress={this.uploadLocations}>
+            <Text style={styles.sectionText}>{I18n.t('settings/push_locations')}</Text>
+          </TouchableOpacity>
+        </View>
+        <View style={[styles.section, styles.first]}>
+          { uploading && <Text style={styles.sectionText}>{I18n.t('locations/uploading')}{`: ${numberUploading - stats.offline}/${numberUploading}`}</Text> }
+          { !uploading && <Text style={styles.sectionText}>{I18n.t('locations/offline')}{`: ${stats.offline + stats.failed}`}</Text> }
+        </View>
+        <View style={styles.section}>
+          <Text style={styles.sectionText}>{I18n.t('locations/successful_upload')}{`: ${stats.uploaded}`}</Text>
+        </View>
+        <View style={styles.section}>
+          <Icon size="medium" family="entypo" name={iconName} colour={colours.black} />
+          <TouchableOpacity onPress={this.onFailedPressed}>
+            <Text style={styles.sectionText}>{I18n.t('locations/Failed_upload')}{`: ${stats.failed}`}</Text>
+          </TouchableOpacity>
+        </View>
+        {stats.failed > 0 &&
+          <CollapsibleList
+            data={failedLocations}
+            open={failedOpen}
+          />
+        }
+      </View>
+    );
+  }
+}
+
+LocationData.propTypes = {
+  failedLocations: PropTypes.array,
+  showPushDialog: PropTypes.func.isRequired,
+  stats: PropTypes.shape({
+    pending: PropTypes.number.isRequired,
+    offline: PropTypes.number.isRequired,
+    failed: PropTypes.number.isRequired,
+    uploaded: PropTypes.number.isRequired,
+  }).isRequired,
+  uploading: PropTypes.bool.isRequired,
+};
+
+LocationData.defaultProps = {
+  failedLocations: [],
+};
+
+const mapStateToProps = (state) => {
+  let failedLocations = {};
+  const stats = {
+    [UploadStatus.pending]: 0,
+    [UploadStatus.offline]: 0,
+    [UploadStatus.failed]: 0,
+    [UploadStatus.uploaded]: 0,
+  };
+
+  Object.entries(state.uploads).forEach(([key, v]) => {
+    stats[v] += 1;
+    if (v === UploadStatus.failed && state.locations[key]) {
+      failedLocations[key] = state.locations[key];
+    }
+  });
+
+  failedLocations = Object.values(failedLocations);
+
+  return {
+    connected: state.connected,
+    stats,
+    failedLocations,
+    uploading: state.uploads.uploading,
+  };
+};
+
+const mapDispatchToProps = dispatch => ({
+  pushLocations: () => dispatch(pushLocalLocations()),
+  requestPushPermission: () => dispatch(requestPushPermission()),
+});
+
+const mergeProps = (stateProps, dispatchProps) => {
+  const props = Object.assign({}, stateProps, dispatchProps);
+
+  props.showPushDialog = () => {
```

Comment:
I wonder if we should move this over to a real thunk so we can isolate this logic if we choose to move it around.

Reply:
This PR or separate?

Reply:
Probably separate, together if you do the feature flag now so just we can keep the code the same. Dealer's Choice.

Reply:
Did in this PR so that I could feature flag